examples/alpaka/pic/pic.cpp

```diff
@@ -0,0 +1,960 @@
+LLAMA_FN_HOST_ACC_INLINE auto squaredNorm(const Vec3F& vec) -> Real
+{
+    return vec(X{}) * vec(X{}) + vec(Y{}) * vec(Y{}) + vec(Z{}) * vec(Z{});
+}
+
+template<typename Vec3F>
+LLAMA_FN_HOST_ACC_INLINE auto cross(const Vec3F& a, const Vec3F& b)
+{
+    return makeV3Real(
+        a(Y{}) * b(Z{}) - a(Z{}) * b(Y{}),
+        a(Z{}) * b(Z{}) - a(Z{}) * b(Z{}),
+        a(Z{}) * b(Y{}) - a(Y{}) * b(Z{}));
+}
+
+template<typename VirtualParticle, typename FieldView>
+LLAMA_FN_HOST_ACC_INLINE void advance_particle(VirtualParticle part, const FieldView& E, const FieldView& B)
+{
+    /* Interpolate fields (nearest-neighbor) */
+    const auto i = static_cast<size_t>(part(R{}, X{}) * inv_dx);
+    const auto i_shift = static_cast<size_t>(part(R{}, X{}) * inv_dx + 0.5);
+    const auto j = static_cast<size_t>(part(R{}, Y{}) * inv_dx);
+    const auto j_shift = static_cast<size_t>(part(R{}, Y{}) * inv_dx + 0.5);
+
+    const auto E_ngp = makeV3Real(E(i_shift, j)(X{}), E(i, j_shift)(Y{}), E(i, j)(Z{}));
+    const auto B_ngp = makeV3Real(B(i, j_shift)(X{}), B(i_shift, j)(Y{}), B(i_shift, j_shift)(Z{}));
+
+    /* q/m [SI] -> 2*PI*q/m in dimensionless */
+    const Real qmpidt = part(Q{}) / part(M{}) * PI * dt;
+    const Real igamma = 1 / sqrt(1 + squaredNorm(part(U{})));
+
+    const auto F_E = qmpidt * E_ngp;
+    const auto F_B = igamma * qmpidt * B_ngp;
+
+    /* Buneman-Boris scheme */
+    const auto u_m = part(U{}) + F_E;
+    const auto u_0 = u_m + cross(u_m, F_B);
+    const auto u_p = u_m + 2 / (1 + squaredNorm(F_B)) * cross(u_0, F_B);
+    const auto u = u_p + F_E;
+
+    /* Update stored velocity and advance particle */
+    const Real igamma2 = 1 / sqrt(1 + squaredNorm(part(U{})));
+    part(U{}) = u;
+    part(R{}) += dt * igamma2 * u;
+    part(R{}, Z{}) = 0;
+
+#ifndef ALPAKA_ACC_GPU_CUDA_ENABLED
+    assert(part(R{}, X{}) >= 0 && part(R{}, Y{}) <= (CELLS_X + 1) * dx);
+    assert(part(R{}, Y{}) >= 0 && part(R{}, Y{}) <= (CELLS_Y + 1) * dx);
+#endif
+}
+
+template<typename VirtualParticle>
+LLAMA_FN_HOST_ACC_INLINE void particle_boundary_conditions(VirtualParticle p)
+{
+    //* Periodic boundary condition
+    if(p(R{}, X{}) < dx)
+        p(R{}, X{}) += (X_ - 3) * dx;
+    else if(p(R{}, X{}) > (X_ - 2) * dx)
+        p(R{}, X{}) -= (X_ - 3) * dx;
+    if(p(R{}, Y{}) < dx)
+        p(R{}, Y{}) += (Y_ - 3) * dx;
+    else if(p(R{}, Y{}) > (Y_ - 2) * dx)
+        p(R{}, Y{}) -= (Y_ - 3) * dx;
+}
+
+template<typename VirtualParticle, typename FieldView>
+LLAMA_FN_HOST_ACC_INLINE void deposit_current(const VirtualParticle& part, FieldView& J)
+{
+    /* Calculation of previous position */
+    const Real igamma = 1 / sqrt(1 + squaredNorm(part(U{})));
+    const auto r_old = part(R{}) - igamma * dt * part(U{}); // Position should be a 2D vector
+
+    /* Particle position to cell coordinates */
+    const auto i = static_cast<size_t>(part(R{}, X{}) * inv_dx);
+    const auto i_shift = static_cast<size_t>(part(R{}, X{}) * inv_dx + 0.5);
```

Comment:
@sliwowitz please figure out whether the `+0.5` and cast is correct. clang-tidy flags this and suggests `lround` if rouding `part(R{}, X{}) * inv_dx` to the nearest integer is intended.

Reply:
I replaced this by `lround` now.